packages/@glimmer/component/addon/-private/component.ts

```diff
@@ -10,12 +10,66 @@ export function setDestroyed(component: GlimmerComponent<object>): void {
   DESTROYED.set(component, true);
 }
 
-export let ARGS_SET: WeakMap<object, boolean>;
+// This provides a type-safe `WeakMap`: the getter and setter link the key to a
+// specific value. This is how `WeakMap`s actually behave, but the TS type
+// system does not (yet!) have a good way to capture that for types like
+// `WeakMap` where the type is generic over another generic type (here, `Args`).
+interface ArgsSetMap extends WeakMap<Args<unknown>, boolean> {
+  get<S>(key: Args<S>): boolean | undefined;
+  set<S>(key: Args<S>, value: boolean): this;
+}
+
+// SAFETY: this only holds because we *only* acces this when `DEBUG` is `true`.
+// There is not a great way to connect that data in TS at present.
+export let ARGS_SET: ArgsSetMap;
 
 if (DEBUG) {
-  ARGS_SET = new WeakMap();
+  ARGS_SET = new WeakMap() as ArgsSetMap;
 }
 
+// --- Type utilities for component signatures --- //
+
+// This provides us a way to have a "fallback" which represents an empty object,
+// without the downsides of how TS treats `{}`. Specifically: this will
+// correctly leverage "excess property checking" so that, given a component
+// which has no named args, if someone invokes it with any named args, they will
+// get a type error.
+declare const Empty: unique symbol;
+type EmptyObject = { [Empty]?: true };
+
+type GetOrElse<Obj, K, Fallback> = K extends keyof Obj ? Obj[K] : Fallback;
+
+/** Given a signature `S`, get back the `Args` type. */
+type ArgsFor<S> = 'Args' extends keyof S
+  ? S['Args'] extends { Named?: object; Positional?: unknown[] } // Are they longhand already?
+    ? {
+        Named: GetOrElse<S['Args'], 'Named', EmptyObject>;
+        Positional: GetOrElse<S['Args'], 'Positional', []>;
+      }
+    : { Named: S['Args']; Positional: [] }
+  : { Named: EmptyObject; Positional: [] };
+
+/** Given any allowed shorthand form of a signature, desugars it to its full expanded type */
+type ExpandSignature<T> = {
```

Comment:
I wonder if there's value (or rather, if the value outweighs the hazard) in exporting some or all of these types as `@internal`.

Otherwise we'll have to duplicate them in Glint, which certainly isn't the end of the world, but there's a certain irony to the fact that this package doesn't even _do_ anything with `Element` or `Blocks` and yet contains the canonical expansion logic for them 😄

Reply:
Exporting as `@internal` makes a lot of sense to me. I had the same thought when authoring it.

Reply:
+1